src/ddmd/dcache.d

```diff
@@ -0,0 +1,377 @@
+ * - better hash map algorithm (it's really dumb)
+ * - better block management (does a linear pass to clean up entries ATM)
+ */
+module ddmd.dcache;
+
+import core.sys.posix.sys.mman;
+import core.sys.posix.fcntl;
+import core.sys.posix.semaphore;
+import core.sys.posix.time;
+import core.sys.posix.unistd;
+import core.stdc.errno;
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+import core.atomic;
+
+version(Posix)
+{
+        
+    struct DCache
+    {
+        /// Load the cache identified by 'keyspace'
+        void initialize(const(char)[] keyspace)
+        {
+            char[512] buf;
+            const(char)* home = getenv("HOME");
+            if (home == null) home = "/tmp";
+            snprintf(buf.ptr, buf.length-1, "%s/.dcache-%s", home, keyspace.ptr);
+            printf("%s\n", buf.ptr);
+            fd = open(buf.ptr, O_CREAT | O_RDWR, S_IRWXU);
+            if (fd < 0) perror("Failed to open DCache mapping");
+            if (posix_fallocate(fd, 0, DATA_SIZE + TOC_SIZE + META_SIZE) < 0) perror("Failed to fallocate DCache mapping");
+            
+            mapping = mapRange(META_SIZE + DATA_SIZE + TOC_SIZE, fd);
+            owner = cast(shared(pid_t)*)mapping.ptr;
+            toc = cast(TocEntry[])mapping[META_SIZE .. META_SIZE + TOC_SIZE];
+            data = cast(Data*)(mapping.ptr + META_SIZE + TOC_SIZE);
+            if (toc.length & (toc.length-1)) abort();
+            ourPid = getpid();
+            // Lock before modification of shared data
+            {
+                lock();
+                scope(exit) unlock();
+                // If first entry is all zeros - new mapping
+                if (data.occupied == 0 && data.size == 0)
+                {
+                    data.size = DATA_SIZE;
+                }
+            }
+            // Anything older then tsThreshold is considered ripe for eviction
+            tsCompilation = time(null);
+            tsThreshold = tsCompilation - 24*3600;
+        }
+
+        /// Clean the cache
+        void truncate()
+        {
+            lock();
+            scope(exit) unlock();
+            (cast(ubyte[])mapping)[]  = 0;
+            data.size = DATA_SIZE;
+        }
+
+        static struct TocEntry
+        {
+            uint hash32;
+            uint index;
+            ulong lastUsed; // seconds since 2000-01-01, == 0 - means entry is empty
+        }
+        static assert(TocEntry.sizeof == 16);
+
+        static struct Data
+        {
+            uint occupied; // 0 - free block, 1 - occupied
+            uint size; // size of block, that includes size of this header
+            uint next; // index from the start of data segement
+            char[0] key; // key data, zero-terminated
+            // ubyte[0] - data
+        }
+        static assert(Data.sizeof == 12);
+
+        string get(const(char)[] key)
+        {
+            lock();
+            scope(exit) unlock();
+            uint h = fnv1a(key);
```

Comment:
We have a better string hash in dmd already.

Reply:
What digestion algorithm is chosen for strings?

Reply:
Murmurhash iirc